rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,752 @@
+        entry = merge.routing_table[i]
+        key, mask = entry.key, entry.mask
+        if any(intersect(key, mask, other.key, other.mask) for other in
+               merge.routing_table[i+1:merge.insertion_index]):
+            # The entry would be partially or wholly covered by another entry,
+            # remove it from the merge and return a new merge.
+            merge = Merge(merge.routing_table, merge.entries - {i})
+            changed = True
+
+            # Check if the merge is sufficiently good
+            if merge.goodness <= min_goodness:
+                merge = Merge(merge.routing_table)  # Replace with empty merge
+                break
+
+    # Return the final merge
+    return merge, changed
+
+
+def refine_downcheck(merge, aliases, min_goodness):
+    """Prune the merge to avoid it covering up any entries which are below the
+    merge insertion position.
+
+    For example, in the (non-orthogonal) table::
+
+        00001 -> N S
+        00011 -> N S
+        00100 -> N S
+        00X00 -> N S
+        XX1XX -> 3 5
+
+    Merging the first four entries would generate the new key-mask ``00XXX``
+    which would be inserted above the entry with the key-mask ``XX1XX``.
+    However ``00XXX`` would stop the key ``00110`` from reaching its correct
+    route, that is ``00110`` would be covered by ``00XXX``. To avoid this one
+    could just abandon the merge entirely, but a better solution is to attempt
+    to reduce the merge such that it no longer covers any entries below it.
+
+    To do this we first identify the bits that ARE ``X`` s in the merged
+    key-mask but which are NOT ``X`` s in the entry that we're covering. For
+    this example this is the 3rd bit. We then look to remove from the merge any
+    entries which are either ``X`` s in this position OR have the same value as
+    in this bit as the aliased entry. As the 4th entry in the table has an
+    ``X`` in this position we remove it, and as the 3rd entry has a ``1`` we
+    also remove it.  For this example we would then consider merging only the
+    first two entries, leading to a new key-mask pair of ``000X1`` which can be
+    safely inserted between ``00X00`` and ``XX1XX``::
+
+        00100 -> N S
+        00X00 -> N S
+        000X1 -> N S
+        XX1XX -> 3 5
+
+    Returns
+    -------
+    :py:class:`~.Merge`
+        New merge with entries possibly removed. If the goodness of the merge
+        ever drops below `min_goodness` then an empty merge will be returned.
+    """
+    # Operation
+    # ---------
+    # While the merge is still better than `min_goodness` we determine which
+    # entries below it in the table it covers. For each of these covered
+    # entries we find which bits are Xs in the merged entry and are NOT Xs in
+    # the covered entry.
+    #
+    # For example:
+    #
+    #     Merged entry:      ...0XXX1...
+    #     Covered entry:     ...010XX...
```

Comment:
010XX could be part of some other entry (as in, XXXXX could contain it, so the aliases dict would have `{XXXXX: {010XXX}}`. I'll point that out in the comment.